File: src/main/java/org/starlisp/core/LispSubr.java
package org.starlisp.core;

abstract class LispSubr extends Procedure {             // FIXME: This class does nothing but modify toString, remove in cleanup later on?

  public LispSubr(String name) {
    super(name);
  }

  public LispSubr(String name, int numArgs) {
    super(name, numArgs);
  }

  public LispSubr(String name, int minArgs, int maxArgs) {
    super(name, minArgs, maxArgs);
  }

  public String toString() {
    return "#<subr " + name + ">";
  }
}

File: src/main/scala/org/starlisp/core/LispSubr.scala
package org.starlisp.core

import scala.Array

object LispSubr {
  def apply(name : String = "", minArgs: Int = 0, maxArgs: Int = Integer.MAX_VALUE)(t:(Array[LispObject]) => LispObject) : LispObject = {
    def f(arr: Array[LispObject]): LispObject = t(arr)
    new LispSubrImplicit(name, minArgs, maxArgs)(f)
  }
}

abstract class LispSubr(name : String = "", minArgs: Int = 0, maxArgs: Int = Integer.MAX_VALUE) extends Procedure(name, minArgs, maxArgs) {
  override def toString: String = "#<subr %s >".format(name)
}

class LispSubrImplicit(name : String = "", minArgs: Int = 0, maxArgs: Int = Integer.MAX_VALUE)(t:(Array[LispObject]) => LispObject) extends Procedure(name, minArgs, maxArgs) {
  override def toString: String = "#<subr %s >".format(name)

  def bar = false

  def apply(objects: Array[LispObject]) : LispObject = t(objects)
}

